File: Linked List/Java/5. Delete a Node in a Linked List.java
/*

 5. Delete a Node in a Linked List

 This problem is to delete a node in a linked list given the value of the node you want to delete.
 It contains a LinkedListNode class to build the linked list, the class LinkedList is the overall list,
 the method addNode from LinkedList is to build up the linked list to test, and to delete a linked list node is in
 the deleteNode method

 Steps to run it correctly:

 First, call instance of LinkedList
 next, call addNode method as many times as needed
 print list before deleting to check changes
 delete node, and print list again to see changes

 Algorithm works by checking if value of deleted node is at front, then traversing through node until either found or at
 end of the list (no more values ahead) where node will not exist

 Time Complexity: O(N)
 Space Complexity O(1) or O(N)
 */



public class LinkedList {
    // front of linked list - head
    LinkedListNode head;

    // LinkedListNode class to build up list (value of node and next if it has a node ahead of it)
    class LinkedListNode {
        int val;
        LinkedListNode next;

        LinkedListNode(int val) {
            this.val = val;
            this.next = null;
        }
    }

    // method to add node to linked list
    public void addNode(int val) {

        LinkedListNode new_node = new LinkedListNode(val);

        if (head == null) {
            head = new LinkedListNode(val);
            return;
        }

        new_node.next = null;

        LinkedListNode pointer = head;

        while (pointer.next != null) {
            pointer = pointer.next;
        }

        pointer.next = new_node;
    }

    // method to delete node from linked list
    public void deleteNode(int val) {
        LinkedListNode pointer = head;

        if (pointer != null && pointer.val == val) {
            head = pointer.next;
            return;
        }

        LinkedListNode prev = null;

        while (pointer != null && pointer.val != val) {
            prev = pointer;
            pointer = pointer.next;
        }

        if (pointer == null) return;

        prev.next = pointer.next;
    }

    // method to print all values in linked list
    public void printList() {
        LinkedListNode temp = head;

        while (temp != null) {
            System.out.println(temp.val);
            temp = temp.next;
        }
    }

    // main method to run test cases
    public static void main(String[] args){
        // create a new linked list object
        LinkedList linkedList = new LinkedList();

        for (int i = 1; i <= Math.pow(10, 3); i++) {
            linkedList.addNode(i);
        }
        System.out.println("BEFORE DELETION");
        linkedList.printList();
        linkedList.deleteNode(999);

        System.out.println("AFTER DELETION");
        linkedList.printList();
    }
}

File: Linked List/C++/5_Delete_a_Node_in_a_Linked_List.cpp
// C++ program to delete a given node in a Linked List

// Code:
#include <bits/stdc++.h>
using namespace std;
/* structured a class of a linked list node */
class Node
{
public:
    int data;
    Node *next;
};

//function to delete the node
void deleteNode(Node *head, Node *n)
{
    // if head node is to be deleted 
    if (head == n)
    {
        if (head->next == NULL)
        {
            cout << "There is only one node."<< " The list can't be made empty ";
            return;
        }
        // Copy the data of next node to head 
        head->data = head->next->data;
        // store address of next node
        n = head->next;
        // Remove the link of next node and connecting to the next element
        head->next = head->next->next;        
        free(n);
        return;
    }    
    
    Node *prev = head;  // to find the previous node
    while (prev->next != NULL && prev->next != n)
        prev = prev->next;    
    if (prev->next == NULL) // Check if node exist
    {
        cout << "\nGiven node is not present in Linked List";
        return;
    }    
    prev->next = prev->next->next; // Remove node from Linked List    
    free(n);
    return;
}
void pushele(Node **head_ref, int new_data) //function to insert a node at the beginning
{
    Node *new_node = new Node();
    new_node->data = new_data;
    new_node->next = *head_ref;
    *head_ref = new_node;
}
void printList(Node *head) //function to print a linked list 
{
    while (head != NULL)
    {
        cout << head->data << " ";
        head = head->next;
    }
    cout << endl;
}
int main()
{
    Node *head = NULL;
    pushele(&head, 1);
    pushele(&head, 8);
    pushele(&head, 23);
    pushele(&head, 7);
    pushele(&head, 9);
    pushele(&head, 4);
    pushele(&head, 5);
    pushele(&head, 0);

    cout << "Given Linked List: ";
    printList(head);    
    cout << "Deleting node " << head->next->next->data << " "; //delete the node 
    deleteNode(head, head->next->next);
    cout << "\nNew Linked List: ";
    printList(head);

    return 0;
}

// Test Case 1:
// Given Linked List: 0 5 4 9 7 23 8 1 
// Deleting node 4 
// New Linked List: 0 5 9 7 23 8 1

// Test Case 2:
// Given Linked List: 0 5 4 9 7 23 8 1 
// Deleting node 9 
// New Linked List: 0 5 4 7 23 8 1 
